Vanja Paunovic - first 40 solutions

dzimiks opened the pull request.

dzimiks committed:
Test commit

dzimiks committed:
Add first 40 solutions

develo-pera commented on tasks/10_remove-whitespaces/remove-whitespaces.js:
```diff
@@ -0,0 +1,3 @@
+export const trimWhiteSpaces = (string) => {
+  return string.replace(/\s/g, '');
```
I like this solution a lot, especially if you came to it on your own (not by googling this for example https://flaviocopes.com/how-to-replace-whitespace-javascript/)

dzimiks replied:
There's a lot of ways to do such a thing. I wanted to practice with RegEx because I used it earlier. Also, I found this website really helpful - https://regex101.com.

develo-pera commented on tasks/14_clone-arr/clone-arr.js:
```diff
@@ -0,0 +1,3 @@
+export const cloneArr = (arr) => {
+  return [...arr];
```
A clever solution, completely valid, but please write this task without using the spread operator. Also, in this case, you could write `const cloneArr = (arr) => [...arr]`

dzimiks replied:
I solved it iteratively.

develo-pera commented on tasks/15_confirm-ending/confirm-ending.spec.js:
```diff
@@ -28,4 +28,4 @@ xdescribe('confirm-ending', () => {
     test('confirmEnding("If you want to save our world, you must hurry. We dont know how much longer we can withstand the nothing", "mountain") should return false', () => {
         expect(confirmEnding("If you want to save our world, you must hurry. We dont know how much longer we can withstand the nothing", "mountain")).toBe(false);
     });
-});
```
This file somehow slipped into your commit without actually being changed. In the future, you should clean cases like this before committing so there's no confusion when someone looks in the history of what was changed.

You can do that by running `git checkout --` command with the path to the file or the folder if you want to revert changes in all files inside that folder.

Example: `git checkout -- tasks/15_confirm-ending/confirm-ending.spec.js` 

This will only work for unstaged files, if you've already done `git add` for that file you need to unstage it first with `git reset HEAD <file>`

dzimiks replied:
This is really useful. The problem was with **EditorConfig** - it has `insert_final_newline = true`, but all test files **don't** have an empty line at the end of a file. So, every time I tried to test my solution, it adds an empty line at the end even if I undo my changes inside the test file.

develo-pera commented on tasks/6_reverse-string/reverse-string.spec.js:
```diff
@@ -16,4 +16,4 @@ xdescribe('reverse-string', () => {
     test('reverseString("Greetings from Earth") should return "htraE morf sgniteerG"', () => {
         expect(reverseString('Greetings from Earth')).toBe('htraE morf sgniteerG');
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/4_stand-in-line/stand-in-line.spec.js:
```diff
@@ -15,9 +15,9 @@ xdescribe('stand-in-line', () => {
+
         expect(testArr[4]).toBe(10);
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/40_find-friend/find-friend.spec.js:
```diff
@@ -31,4 +31,4 @@ xdescribe('find-friend', () => {
     test(`findFriend(['blue','red','blue','red','red','blue','red']) should return 1`, () => {
         expect(findFriend(['blue','red','blue','red','red','blue','red'])).toBe(1);
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/3_is-boolean/is-boolean.spec.js:
```diff
@@ -41,4 +41,4 @@ xdescribe('is-boolean', () => {
     test('isBoolean("false") should return false', () => {
         expect(isBoolean("false")).toBe(false);
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/34_max-triplet-sum/max-triplet-sum.spec.js:
```diff
@@ -12,4 +12,4 @@ xdescribe('max-triplet-sum', () => {
     test(`maxTripletSum([3,2,6,8,2,3]) should return 17`, () => {
         expect(maxTripletSum([3,2,6,8,2,3])).toBe(17);
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/29_remove-duplicates/remove-duplicates.spec.js:
```diff
@@ -20,4 +20,4 @@ xdescribe('remove-duplicates', () => {
     test(`removeDuplicates([2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1]) should return [2, 1]`, () => {
         expect(removeDuplicates([2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1])).toEqual([2, 1]);
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/25_rot13-encoder/rot13-encoder.spec.js:
```diff
@@ -16,4 +16,4 @@ xdescribe('rot13-encoder', () => {
     test('rot13Encoder("Hello World!") should encode to "Uryyb Jbeyq!"', () => {
         expect(rot13Encoder("LOVE?")).toBe('YBIR?');
     });
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/1_golf-score/golf-score.spec.js:
```diff
@@ -46,4 +46,4 @@ xdescribe('golf-score', () => {
         expect(golfScore(5, 9)).toBe(GOLF_SCORES.GO_HOME);
     });
 
-});
```
Same as [tasks/15_confirm-ending/confirm-ending.spec.js](https://github.com/maleksandar/gd-ui-internship-2020-100-problems/pull/7/files#diff-c5513d47cd98222e3767cd6863062dc2)

develo-pera commented on tasks/16_repeat-string/repeat-string.js:
```diff
@@ -0,0 +1,3 @@
+export const repeatStringNumTimes = (string, numTimes) => {
+  return (numTimes < 0) ? '' : string.repeat(numTimes);
```
Parentheses around `(numTimes < 0)` are unnecessary in this case. Also, what if the string is empty, should you call repeat?

dzimiks replied:
I fixed it with `(numTimes < 0 || string === '')`.

develo-pera commented on tasks/17_truncate-string/truncate-string.js:
```diff
@@ -0,0 +1,9 @@
+export const truncateString = (string, length) => {
+  if (length <= 3) {
+    return string.slice(0, length).concat('...');
```
This could be done with template strings also return ``` return `${string.slice(0, length)}...` ```

dzimiks replied:
Great solution! :)

develo-pera commented on tasks/17_truncate-string/truncate-string.js:
```diff
@@ -0,0 +1,9 @@
+  if (length <= 3) {
+    return string.slice(0, length).concat('...');
+  } else if (length < string.length) {
+    return string.slice(0, length - 3).concat('...');
```
Same here

develo-pera commented on tasks/18_humanize-format/humanize-format.js:
```diff
@@ -0,0 +1,18 @@
+export const humanizeFormat = (...args) => {
```
Why would you use rest parameter here? The task was very clear, "write a JavaScript function to humanized **number**", function signature should tell me very clearly that I need to pass a number, just one number and nothing more.

dzimiks replied:
I used rest operator because of the first test case - `expect(humanizeFormat()).toBe('')`. Since this case doesn't include any argument, I checked `args.length === 0` to see if there is an argument passed to the function.

develo-pera commented on tasks/18_humanize-format/humanize-format.js:
```diff
@@ -0,0 +1,18 @@
+  }
+
+  const number = args[0].toString();
+  const lastDigit = number.slice(-1);
```
Does this solution covers all edge cases, what about 11, 12, 13, 111, 113? Those numbers should have suffixes 11th, 12th, 13th, 111th, 113th

dzimiks replied:
My solution was incorrect. I forgot to check those edge cases. It's fixed now.

develo-pera commented on tasks/19_array-chunks/array-chunks.js:
```diff
@@ -0,0 +1,10 @@
+export const arrayChunks = (arr, size) => {
+  let res = [];
+  const n = arr.length;
```
This const is unnecessary, `array.length` can go inside for loop condition `for (let i = 0; i < array.length; i += size)`

develo-pera commented on tasks/19_array-chunks/array-chunks.js:
```diff
@@ -0,0 +1,10 @@
+  let res = [];
+  const n = arr.length;
+
+  for (let i = 0; i < n; i += size) {
```
If someone passes 0 or a negative number as a second argument we will end up in the indefinite loop, and we should always prevent that.

dzimiks replied:
I didn't cover that corner case. It's fixed now.

develo-pera commented on tasks/20_truncate-array/truncate-array.js:
```diff
@@ -0,0 +1,3 @@
+export const truncateArray = (arr, n) => {
+  return arr.splice(n);
```
This will change the original array. We should tend to write pure functions, without side effects and without unwanted mutations.

dzimiks replied:
I agree. I didn't think about that when I used `splice()` in this task. It's fixed now.

develo-pera commented on tasks/22_remove-falsy-values/remove-falsy-values.js:
```diff
@@ -0,0 +1,4 @@
+export const removeFalsyValues = (arr) => {
+  return arr.filter((elem) => ![false, null, 0, NaN, undefined, '']
```
Since you already use the built-in filter function you could just do `removeFalsyValues = (arr) => arr.filter(elem => elem)`. Only truthy values will be included in array after filtering with that function.

dzimiks replied:
I agree. I even tried with `arr.filter((elem) => !!elem)`, which is also correct, but it is unnecessary use of **double negation**.

develo-pera commented on tasks/27_basic-subclasses/basic-subclasses.js:
```diff
@@ -0,0 +1,45 @@
+export class God {
```
Can you write this using constructor functions and not ES6 classes?

dzimiks replied:
I've changed it to constructor functions.

develo-pera commented on tasks/3_is-boolean/is-boolean.js:
```diff
@@ -0,0 +1,3 @@
+export const isBoolean = (value) => {
+  return typeof value === 'boolean';
```
Can be shortened to `const isBoolean = value => typeof value === 'boolean';`.

Fun fact, if for some crazy reason there's a boolean created using constructor `new Boolean`, for example `const isTrue = new Boolean(true)` this check won't catch it because typeof will return `'object'` for it.

dzimiks replied:
Yes, that's true. If we are using constructor, it creates an object, but not boolean as a primitive type. Technically, my solution is correct, but I edited it to work with the constructor `new Boolean` as well.

develo-pera commented on tasks/8_factorize-number/factorize-number.js:
```diff
@@ -0,0 +1,9 @@
+export const factorialize = (n) => {
+  let fact = 1;
+
+  for (let i = 2; i <= n; i++) {
```
The completely valid and correct solution, but you could also use recursion here instead of going iteratively, just as exercise

dzimiks replied:
I've done it recursively as well.

develo-pera commented on tasks/28_vowel-count/vowel-count.js:
```diff
@@ -0,0 +1,5 @@
+export const getVowelCount = (string) => {
+  return [...string]
+      .filter((char) => ['a', 'e', 'i', 'o', 'u']
+          .includes(char.toLowerCase())).length;
```
```suggestion
     .includes(char.toLowerCase())).length;
```

develo-pera replied:
it's more readable if you remove indentation here

dzimiks replied:
**ESLint** doesn't allow me to indent it like that. I'll refactor this part of the code.

develo-pera commented on tasks/2_convert-to-f/convert-to-f.js:
```diff
@@ -0,0 +1,3 @@
+export const convertToF = (x) => {
+  return x * 9 / 5 + 32;
```
could be shortened to `const convertToF = x =>  x * 9 / 5 + 32;`

develo-pera commented on tasks/9_find-numbers/find-numbers.js:
```diff
@@ -0,0 +1,3 @@
+export const getNumbers = (string) => {
+  return string.match(/\d+/g).map((item) => parseInt(item));
```
Note: this is ok because we are sure that we will have an array of chars that can be converted to numbers, but `parseInt` isn't always the best solution for string to number conversion. I encourage you to do research and see difference between `parseInt` and explicit conversion with `Number()`

dzimiks replied:
That's really interesting. Function **parseInt(string)** will convert a string containing non-numeric characters to a number, as long as the string begins with numeric characters, but function **Number(string)** will return **NaN** if the string contains any non-numeric characters. Since I used **RegEx** to match all numbers with `\d+`, this will work for all integer values. Actually, I can write it even shorter: `return string.match(/\d+/g).map(Number)`.

develo-pera commented on tasks/32_numbers-to-reversed-array/numbers-to-reversed-array.js:
```diff
@@ -0,0 +1,3 @@
+export const numberToReversedArray = (number) => {
+  return Array.from(number.toString().split('').reverse().join('')).map(Number);
```
`number.toString().split('')` already creates array, joining and creating array with `Array.from` is unnecessary

dzimiks replied:
That's pretty accurate. I added unnecessary code.

develo-pera commented on tasks/7_random-fraction/random-fraction.js:
```diff
@@ -0,0 +1,3 @@
+export const randomFraction = () => {
+  return Math.random() % 1;
```
Can you explain what is the purpose of % 1?

dzimiks replied:
It was completely unnecessary since Math.random() already returns a number between 0 and 1.

develo-pera commented on tasks/31_reverse-string-in-range/reverse-string-in-range.js:
```diff
@@ -0,0 +1,6 @@
+  const start = string.substring(0, range[0]);
+  const mid = string.substring(range[0], range[1] + 1);
+  const end = string.substring(range[1] + 1, string.length);
+  return start.concat(mid.split('').reverse().join('')).concat(end);
```
Using template string here would probably improve readability

dzimiks replied:
I refactored code to look like this:

```
const start = string.substring(0, range[0]);
const mid = string.substring(range[0], range[1] + 1);
const midReversed = mid.split('').reverse().join('');
const end = string.substring(range[1] + 1, string.length);
```

I think it's weird to have **\`${start}${midReversed}${end}\`** as a return statement. I can return it as `start.concat(midReversed).concat(end)` or even `[...start, ...midReversed, ...end].join('')`. I prefer this solution: **`start.concat(midReversed).concat(end)`**.

develo-pera commented on tasks/33_array-plus-array/array-plus-array.js:
```diff
@@ -0,0 +1,13 @@
+    res.push(arr1[i] + arr2[i]);
+  }
+
+  arr1 = arr1.splice(n);
```
We should never reassign or change the argument value. Create a new variable using `const` or `let` if you need, but don't change argument, also avoid using splice before cloning array because it will mutate the original array.

dzimiks replied:
I agree. I fixed it.

develo-pera commented on tasks/33_array-plus-array/array-plus-array.js:
```diff
@@ -0,0 +1,13 @@
+export const arrayPlusArray = (arr1, arr2) => {
+  let n = Math.min(arr1.length, arr2.length);
```
Avoid using `n` or single-letter variables, always tend to write variable and function names that clearly represent what they are

dzimiks replied:
My bad. I should be more descriptive about variable names.

develo-pera commented on tasks/36_row-weight/row-weight.js:
```diff
@@ -0,0 +1,15 @@
+export const rowWeights = (arr) => {
+  const n = arr.length;
```
This can be  moved to for loop condition

develo-pera commented on tasks/38_max-product/max-product.js:
```diff
@@ -0,0 +1,15 @@
+export const maxProduct = (arr) => {
+  const n = arr.length;
+  let product = 0;
+  let maxProduct = -Infinity;
```
It's good that you know about `Infinity` but in this case, code would look clearer if you assigned `arr[0] * arr[1]` immediately to `maxProduct`

dzimiks replied:
That's correct. Then I would go from the second index in for loop.

develo-pera commented on tasks/5_profile-lookup/profile-lookup.js:
```diff
@@ -0,0 +1,31 @@
+export const lookupProfile = (firstName, prop) => {
```
What will be the output in this case `lookupProfile('Harry', 'lastName');`?

dzimiks replied:
I fixed this issue.

dzimiks committed:
Refactor code based on code review

- Task 2: Use the shortened expression.
- Task 3: Check for new Boolean.
- Task 5: Refactor incorrect code.
- Task 7: Remove the '%' operator.
- Task 8: Add recursive implementation.
- Task 9: Use Number() instead of parseInt().
- Task 14: Add iterative implementation.
- Task 16: Handle empty string.
- Task 17: Use template literals instead of concat().
- Task 18: Cover edge cases and remove rest operator.
- Task 19: Prevent indefinite loop.
- Task 20: Write pure function.
- Task 22: Use a better solution.
- Task 27: Use constructor functions instead of classes.
- Task 28: Refactor code.
- Task 31: Refactor code to be more readable.
- Task 32: Remove unnecessary code.
- Task 33: Prevent mutating the original array.
- Task 36: Remove unnecessary variable.
- Task 38: Use multiplication of the first two numbers from an array instead of -Infinity value.
